pep-0669.rst

```diff
@@ -0,0 +1,318 @@
+  instrumentation.monitor_on(codeobject, offset)
+
+All functions return ``True`` if a monitor checkpoint was present,
+or ``False`` if a monitor checkpoint was not present.
+Turning on, or off, a non-existent checkpoint is a no-op;
+no exception is raised.
+
+To register a callable for monitoring function events call::
+
+  instrumentation.monitor_register(func)
+
+The callback function will be called with the code object and offset as arguments::
+
+  func(code: CodeType, offset: int)
+
+For optimizing virtual machines, such as future versions of CPython
+(and ``PyPy`` should they choose to support this API), a call to
+``insert_monitors`` and ``remove_monitors`` in a long running program
+could be quite expensive, possibly taking 100s of milliseconds as it
+triggers de-optimizations. Repeated calls to ``insert_monitors``
+and ``remove_monitors``, as may be required in an interactive debugger,
+should be relatively inexpensive.
+
+Combining Checkpoints
+---------------------
+
+Only one instrumentation checkpoint and one monitoring checkpoint is allowed
+per bytecode instruction. It is possible to have both a monitoring and
+instrumentation checkpoint on the same instruction; they are independent.
+Monitors will be called before instrumentation if both are present.
+
+Backwards Compatibility
+=======================
+
+This PEP is fully backwards compatible.
+
+We may seek to remove ``sys.settrace`` in the future once the APIs provided
+by this PEP have been widely adopted, but that is for another PEP.
+
+
+Security Implications
+=====================
+
+Allowing modification of running code has some security implications,
+but no more than the ability to generate and call new code.
+
+All the functions listed above will trigger audit hooks.
+
+
+Implementation
+==============
+
+The implementation of this PEP will be built on top of PEP 659 quickening.
+Instrumentation or monitoring of a code object will cause it to be quickened.
+Checkpoints will then be implemented by inserting one of several special
+``CHECKPOINT`` instructions into the quickened code. These instructions
+will call the registered callable before executing the original instruction.
+
+Note that this can interfere with specialization, which will result in
+performance degradation in addition to the overhead of calling the
+registered callable.
+
+Implementing tools
+==================
+
+It is the philosophy of this PEP that third-party tools should be able to
+achieve high-performance, not that it should be easy for them to do so.
+This PEP provides the necessary API for tools, but does nothing to help 
+them determine when and where to insert instrumentation or monitors.
+
+Debuggers
+---------
+
+Inserting breakpoints
+'''''''''''''''''''''
+
+Breakpoints should be implemented as monitors.
+To insert a breakpoint at a given line, the matching instruction offsets
+should be found from ``codeobject.co_lines()``.
+Then a monitor should be added for each of those offsets.
```

Comment:
I'm not super familiar with co_lines(). When does it return multiple offsets? Only when the code of that line is duplicated (e.g. by loop unrolling or similar optimizations)?